### PR TITLE
feat(IPVC-2231): add script to parse genomic gff 

### PR DESCRIPTION
**Pull request by sptaylor**
This adds a new script for parsing gff. I read from the ticket that we wanted a new script, but if we'd like to replace the existing one, I can do that instead.

I included the shell scripts and changes needed to run these steps just to demonstrate that they work. I figure we'll refactor this so we only pull the database once. The parse gff script runs in the full slice, but fails due to not having upstream step plugged in yet (I believe that is the gbff processing step from [here](https://lucid.app/lucidchart/8e0e2a73-24cb-49a9-9f97-2f26019f822d/edit?invitationId=inv_4ceff0b0-df98-4ea4-8a2d-dbf326096189&page=ceDdGotQlEn_#))

```DETAIL:  Key (tx_ac)=(NM_001089591.2) is not present in table "transcript".```

### CODE REVIEW

File: src/uta/models.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # schema name support
 # also see etc/uta.conf
 
-schema_version = "1.1"
+schema_version = "20210129b"
```
Comment (bsgiles73) [inline src/uta/models.py]: I understand why we changed this here. But I think we will need to leave it, or change it back later. Either we need a default name or an environment variable to set this. I have reasons why I think a default might be better.

File: sbin/ncbi_parse_genomic_gff.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+from typing import List, Optional
+import gzip
+from collections import defaultdict
+from dataclasses import dataclass
+
+import logging
+
+from uta.formats.exonset import ExonSet, ExonSetWriter
+
+logging.basicConfig(level=logging.INFO)
```
Comment (bsgiles73) [inline sbin/ncbi_parse_genomic_gff.py]: I think we should setup logging the same way across existing scripts and new. Existing scripts setup the logger in this way...

```
    logging_conf_fn = pkg_resources.resource_filename(
        "uta", "etc/logging.conf")
    logging.config.fileConfig(logging_conf_fn)
    logging.getLogger().setLevel(logging.INFO)
    logger = logging.getLogger(__name__)
```

Comment (bsgiles73) [reply sbin/ncbi_parse_genomic_gff.py]: https://github.com/biocommons/uta/blob/main/sbin/ncbi-parse-gbff#L126

File: sbin/ncbi_parse_genomic_gff.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    # NC_000001.10	BestRefSeq	exon	12613	12721	.	+	.	ID=exon-NR_046018.2-2;Parent=rna-NR_046018.2;Dbxref=GeneID:100287102,Genbank:NR_046018.2,HGNC:HGNC:37102;gbkey=misc_RNA;gene=DDX11L1;product=DEAD/H-box helicase 11 like 1 (pseudogene);pseudo=true;transcript_id=NR_046018.2
+    # NC_000001.10	BestRefSeq	exon	13221	14409	.	+	.	ID=exon-NR_046018.2-3;Parent=rna-NR_046018.2;Dbxref=GeneID:100287102,Genbank:NR_046018.2,HGNC:HGNC:37102;gbkey=misc_RNA;gene=DDX11L1;product=DEAD/H-box helicase 11 like 1 (pseudogene);pseudo=true;transcript_id=NR_046018.2
+
+    fields = line.strip().split("\t")
+    if len(fields) != 9:
+        raise ValueError(f"Expected 9 tab-separated fields, got {len(fields)}")
+
+    seqid, source, feature, start, end, score, strand, phase, attributes_str = fields
+
+    if feature != "exon" or not seqid.startswith("NC_"):
```
Comment (bsgiles73) [inline sbin/ncbi_parse_genomic_gff.py]: Filtering for alternate accessions to "NC_" is something that I had in my script, but not sure we have a requirement to filter down alignments just to complete chromosome contigs. I think I had it just for testing. I think we should remove it. But open for discussion as I am not sure what purpose they serve.

File: sbin/ncbi_parse_genomic_gff.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    parent_id = attributes.get("parent")
+    transcript_id = attributes.get("transcript_id")
+    if (
+        not transcript_id
+        or (not transcript_id.startswith("NM_") and not transcript_id.startswith("NR_"))
+        or not parent_id
+    ):
+        return
+    try:
+        exon_number = _get_exon_number_from_id(alignment_id=attributes.get("id"))
```
Comment (bsgiles73) [inline sbin/ncbi_parse_genomic_gff.py:82]: Do we need the exon number?

File: sbin/ncbi_parse_genomic_gff.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        for line in f:
+            if line.startswith("#"):
+                continue
+            try:
+                record = parse_gff_record(line)
+            except ValueError as e:
+                raise Exception(f"Failed at line :{line} with error: {e}")
+            if record:
+                tx_data[record.parent_id].append(record)
+    return {k: _sort_exons(v) for k, v in tx_data.items()}
```
Comment (bsgiles73) [inline sbin/ncbi_parse_genomic_gff.py:117]: Having exons here though is nice.

**Review by bsgiles73 — APPROVED** [review]
A few comments here and there put this looks good to me. We can update run scripts and some other values as we go.

I also think we should move old scripts like `sbin/ncbi-parse-gff` to `archive/`.

File: sbin/ncbi_parse_genomic_gff.py (hunk 3 above)
Comment (sptaylor) [reply sbin/ncbi_parse_genomic_gff.py]: I'll remove it

**Pull request merged**